core/storage/trie/impl/polkadot_trie_cursor.cpp

```diff
@@ -0,0 +1,203 @@
+        // find the rightmost child
+        for (int8_t i = branch->kMaxChildren - 1; i >= 0; i--) {
+          if (branch->getChild(i) != nullptr) {
+            OUTCOME_TRY(c, trie_.retrieveChild(branch, i));
+            current = c;
+          }
+        }
+
+      } else {
+        BOOST_ASSERT_MSG(
+            false,
+            "must not reach here, as with using retrieveChild there should "
+            "appear no other trie node type than leaf and both branch types");
+        current_ = nullptr;
+        return outcome::success();
+      }
+    }
+    current_ = current;
+    return outcome::success();
+  }
+
+  bool PolkadotTrieCursor::isValid() const {
+    return current_ != nullptr and current_->value.has_value();
+  }
+
+  outcome::result<void> PolkadotTrieCursor::next() {
+    if (not visited_root_) {
+      current_ = trie_.getRoot();
+    }
+    if (current_ == nullptr) {
+      return trie_.getRoot() == nullptr ? Error::NULL_ROOT
+                                        : Error::INVALID_CURSOR_POSITION;
+    }
+    if (not visited_root_ and trie_.getRoot()->value.has_value()) {
+      visited_root_ = true;
+      return outcome::success();
+    }
+    visited_root_ = true;
+    do {
+      if (current_->getTrieType() == NodeType::Leaf) {
+        if (last_visited_child_.empty()) {
+          current_ = nullptr;
+          return outcome::success();
+        }
+        // assert last_visited_child_.back() == current.parent
+        auto p = last_visited_child_.back().first;  // self.current.parent
+        while (not hasNextChild(p, last_visited_child_.back().second)) {
+          last_visited_child_.pop_back();
+          if (last_visited_child_.empty()) {
+            current_ = nullptr;
+            return outcome::success();
+          }
+          p = last_visited_child_.back().first;  // p.parent
+        }
+        auto i = getNextChildIdx(p, last_visited_child_.back().second);
+        OUTCOME_TRY(c, trie_.retrieveChild(p, i));
+        current_ = c;
+        updateLastVisitedChild(p, i);
+
+      } else if (current_->getTrieType() == NodeType::BranchEmptyValue
+                 or current_->getTrieType() == NodeType::BranchWithValue) {
+        auto p = std::dynamic_pointer_cast<BranchNode>(current_);
+        if (last_visited_child_.empty()
+            or last_visited_child_.back().first != p) {
+          last_visited_child_.emplace_back(std::make_pair(p, -1));
+        }
+        while (not hasNextChild(p, last_visited_child_.back().second)) {
+          last_visited_child_.pop_back();
+          p = last_visited_child_.back().first;  // p.parent
+        }
+        auto i = getNextChildIdx(p, last_visited_child_.back().second);
+        OUTCOME_TRY(c, trie_.retrieveChild(p, i));
+        current_ = c;
+        updateLastVisitedChild(p, i);
+      }
+    } while (not current_->value.has_value());
+  }
+
+  outcome::result<void> PolkadotTrieCursor::prev() {
+    BOOST_ASSERT_MSG(false, "Not implemented");
+  }
```

Comment:
It's okay to leave unimplemented method here? Think about to implement if it needs and not so hard.